aten/src/ATen/SparseTensorImpl.h

```diff
@@ -275,6 +275,9 @@ struct TORCH_API SparseTensorImpl : public TensorImpl {
     AT_ASSERT(new_nnz <= nnz());
     indices_ = indices_.narrow(1, 0, new_nnz);
     values_ = values_.narrow(0, 0, new_nnz);
+    if (new_nnz < 2) {
+      coalesced_ = true;
+    }
```

Comment:
This change appears to be unnecessary because almost all `set_nnz_and_narrow` current usages follow `_coalesced_(...)` method call. The exception is `_sspaddmm_out_cpu` which may require separate analysis when its result is guaranteed to be coalesced.
So, suggest:
```suggestion
```

Reply:
I'm not sure I see much difference between these two changes, so if one is ok I think the other is too: both modify their respective setters to opportunistically set `coalesced_` true when the condition can be easily detected, even though modifying `coalesced_` isn't explicitly called out in the name of the method. 

It's true that many call sites (of both methods) set `coalesced_` themselves, but in the absence of more consistency in how these are called I don't know that that should matter when considering these changes. At any rate the modifications are strictly conservative, in the sense that they preserve the invariant that `indices_` and `values_` are coalesced if (but not only if) `coalesced_` is true. 

If the changes were expensiveI could see that being an argument against them, but I don't think they are.

@pearu I'm ok with these changes, but I know you and @amjames have discussed it at some length - how strong is your objection to them?

Reply:
The objection is not strong. My position is to be conservative in changing code when (i) it is not strictly related to the reported issue and (ii) there appears to be a duplication of initializing `coalesced_`. So, feel free to overrule as the changes ought to be safe.

However, the failing tests need to be addressed anyway and when doing so, we may learn something, or not :)

Reply:
Agreed - I think Andrew's working through the test issues now.

Reply:
I am working on the changes to correct the assumptions causing the tests to fail now. Once we have a look over those diffs it may be enlightening. 

The significant number of the failures come from an assert in a setup function and have nothing to do with the postconditions of an operation. We shall see if that changes as I get further along.